README.md
# GP Bot
###### [1.1.0](docs/release_notes/1.1.0.md)
\
Cansado de **gastar seu tempo** lançando o [TQI-GP](https://helpdesk.tqi.com.br/sso/login.action) **na mão todos os dias**? 

Ou pior, na **correria**, com o **prazo** quase acabando e cheio de **cobranças**?

Deixe o **GP Bot** fazer todo o trabalho **automaticamente** e em instantes!

* faça o lançamento das **horas do dia** ou **de todos os dias do mês de uma só vez**!
* informe **dias com horas diferentes de 8** se houver: ganhe rapidez com precisão!
* escolha os **dias que você quer tratar manualmente** e ajuste você mesmo na página do TQI-GP

E o melhor: levando **menos de 1 minuto!** :O

## Requisitos
- Linux 64-bit;
- Google Chrome entre (e incluindo) as versões 61 e 69;
- Java Runtime Environment 8 ou superior.

## Faça o download

Baixe a última versao [aqui](release/gp-bot.zip?raw=true) e descompacte o arquivo em um diretório de sua preferência.

## Executando o GP Bot

Há duas maneiras de informar os dados de lançamento ao bot:

- via **[console](#lancamentoConsole)**, informando manualmente os parâmetros de execução. Use esta opção se quiser preencher tudo com calma, passo a passo.
- via **[linha de comando](#lancamentoLinhaComando)**. Para quem quer mais **rapidez**! Use também esta opção se necessitar de um **comando único** que possa ser configurado numa **ferramenta de agendamento** como o **cron** ou de **automatização de execução** como o **Jenkins**.

Em comum, as duas opções são iniciadas com a execução do arquivo `gp-bot.jar`:

1. para preenchimento dos dados no console: 
`java -jar gp-bot.jar`

2. para execução com parâmetros via linha de comando: 
`java -jar gp-bot.jar [parametros]`

No caso da execução via **linha de comando** (veja detalhes abaixo), o console pedirá que o usuário complete somente as **informações obrigatórias** que não tiverem sido preenchidas. São **obrigatórios** para o lançamento do TQI-GP:

- usuário (TQI);
- senha;
- aplicativo;
- atividade.

Caso tenham sido **todos** informados via linha de comando o GP Bot iniciará **imediatamente** o processo de lançamento.

## <a name="lancamentoConsole"></a>Lançamento pelo console 
Pelo console, informe em sequência:
- **Usuário**: seu username TQI. Exemplo: _jose.silva_
- **Senha**: sua senha (não se preocupe, ela será ofuscada).
- **Nome do aplicativo**: Aplicativo/projeto em que você está trabalhando. Exemplo: _Rede Acme Fase 3_
- **Atividade**: Será exibida a lista de atividades disponíveis para lançamento no TQI-GP. Informe um código numérico da lista ou se preferir o nome exato da atividade. Exemplo: _1_ ou _Codificação_ para lançar a atividade de codificação.
- **Lançar somente hoje (opcional):** Digite _s_ ou _sim_ se quiser que seja lançado somente o dia corrente, com 8 horas. Caso seja informado, as opções complementares abaixo deixarão de ser exibidas.
- **Mês (opcional)**: o mês (numeral) em que serão feitos os lançamentos. Se não informado, o **GP Bot** **assumirá o mês atual**. Exemplo: _3_ (Março)
- **Dias a serem ignorados (opcional)**: dias, separados por vírgulas, para os quais o **GP Bot não efetuará lançamentos**. Exemplo: _3, 5, 17_
- **Dias com horas específicas (opcional)**: dias para os quais o **GP Bot** deverá **lançar horas diferentes de 8**. Separados por vírgulas, a quantidade de horas deverá ser informada entre parênteses no **formato fracional ou horário**. Exemplo: _3(7), 5(7.5), 17(6:35)_  - sete horas no dia 3, sete horas e meia no dia 5 e seis horas e trinca e cinco minutos no dia 17)
   
## <a name="lancamentoLinhaComando"></a>Lançamento com parâmetros via linha de comando
Pela linha de comando, complete os parâmetros passados ao executável:

`java -jar gp-bot.jar [parametros]`

com as opções abaixo, em qualquer ordem:

- `--usuario`, para seu usuário;
- `--senha`, para sua senha;
- `--aplicativo`, para o nome do aplicativo;
- `--atividade`, para a atividade desempenhada;
- `--hoje`, para efetuar o lançamento somente do dia corrente;
- `--mes`, para o mês desejado;
- `--skipdays`, para os dias a serem ignorados;
- `--customdays`, para os dias com horas específicas.

O formato dos dados é o mesmo descrito na seção de lançamento via console. 

Para parâmetros de lista de valores ou com mais de uma palavra separada por espaços utilize **aspas** no preenchimento. Exemplo: `java -jar gp-bot --aplicativo='Uma app Fantástica'`

### Exemplos de execução com parâmetros via linha de comando

##### 1. Lançamento de todo o mês com atividade de codificação, execução imediata (todos os parâmetros obrigatórios informados):

`java -jar gp-bot.jar --usuario=jose.silva --senha=foobar123 --aplicativo=AppCliente --atividade=1`

##### 2. Lançamento do dia atual com atividade de codificação, execução imediata:

`java -jar gp-bot.jar --usuario=jose.silva --senha=foobar123 --aplicativo=AppCliente --atividade=1 --hoje`

##### 3. Lançamento do dia atual com atividade de codificação, console aberto para preenchimento manual da senha (parâmetro obrigatório não informado):

`java -jar gp-bot.jar --usuario=jose.silva --aplicativo=AppCliente --atividade=1 --hoje`

##### 4. Lançamento de todo o mês de Março exceto os dias  3, 5 e 17, com os dias 25 e 14 customizados com 7h e 7h35min respectivamente, com atividade de codificação e execução imediata:

`java -jar gp-bot.jar --usuario=jose.silva --senha=foobar123 --aplicativo='Rede Acme Fase 3' --atividade='Codificação' --mes=3 --skipdays='3, 5, 17' --customdays='25(7), 14(7:35)'`

release_notes.txt
1. Melhorias na engine

1.1 Com o uso da configuração "headless" do Chrome, o GP Bot por padrão não utiliza mais a interface gráfica do navegador para efetuar os lançamentos.

Para habilitar o lançamento com o uso da GUI do Chrome, informe o parâmetro --visivel na linha de comando.

1.2 É possível agora lançar somente o dia atual:
- via console: preencha a opção correspondente com S ou sim;
- via linha de comando: informe o parâmetro --hoje. 

Exemplo: java -jar gp-bot.jar --usuario=jose.silva --aplicativo=PagSeguro --atividade=Codificação --hoje

1.3 É possível utilizar uma versão customizada do ChromeDriver, diferente daquelas distribuídas com a aplicação (2.34, 2.37 e 2.38). Isso possibilita a atualização do seu Chrome para versões mais novas sem perda de funcionalidade pelo GP Bot.

Para isso, faça o download da versão desejada do ChromeDriver (http://chromedriver.chromium.org/) e cole o arquivo binário "chromedriver" no subdiretório /chromedriver da pasta onde se encontra o executável da aplicação.

1.3 A janela do Chrome é fechada automaticamente ao final do lançamento no caso de execução --visivel

1.4 Na opção de lançamento de horas customizadas são agora suportados dois formatos:
- fracionário: 8(7.5) -> dia 8, sete horas e meia
- horário: 8(7:30) -> dia 8, sete horas e trinta minutos

Ambos podem ser usados de forma equivaçente.

2. Melhorias no console de entrada

2.1 É exibida agora a lista completa de atividades do GP para facilitar o preenchimento. É possível informar a atividade tanto pelo nome exato (como antes) quanto pelo código da lista:
- via console: informe o nome exato da atividade ou o código da lista de atividades;
- via linha de comando: informe o parâmetro --atividade com o nome exato ou o código da lista de atividades.

Exemplos: 
java -jar gp-bot.jar --atividade=Codificação ou
java -jar gp-bot.jar --atividade=1 para lançar a atividade de Codificação

2.2 O console não exige mais o preenchimento de parâmetros já informados via linha de comando, somente os obrigatórios não preenchidos.

3. Melhorias no log

O log está mais informativo.

3.1 Exibe agora os feriados e dias de fim de semana detectados no intervalo de lançamento em questão

3.2 Exibe mensagem de login na intranet TQI.

3.3 Exibe mensagens de leituras de arquivos de configuração da aplicação.

3.3 Exibe mensagens de erro devolvidas pelo TQI-GP no navegador durante o preenchimento.

- aplicativo opcional para teste de sistema



src/main/java/org/mfs/gpbot/GPBotSetup.java
package org.mfs.gpbot;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.text.MessageFormat;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;
import java.util.stream.Stream;

import org.apache.commons.lang3.StringUtils;
import org.apache.log4j.Logger;
import org.mfs.gpbot.enumeration.InputParameterEnum;
import org.openqa.selenium.chrome.ChromeDriver;

/**
 * Captura dados das entradas, identifica a versao do Chrome e do ChromeVersion
 * aplicavel
 *
 * @author Michel Suzigan
 *
 */
public class GPBotSetup {

	private static final Logger LOGGER = Logger.getLogger(GPBotSetup.class);
	private static final String CHROME_VERSION_COMMAND_OUTPUT_PATTERN = "Google Chrome \\d{2}.*";

	private GPBotSetup() {

	}

	public static GPBotData getData(String[] args) {
		GPBotData data;

		try {
			data = getDataFromArgs(args);

			if (!data.hasEssentialInformation()) {
				data = getDataFromDialog(data);
			}

		} catch (Exception e) {
			throw new IllegalArgumentException("Erro ao ler informacoes de entrada", e);
		}

		if (!data.hasEssentialInformation()) {
			throw new IllegalArgumentException(
					"Informacoes insuficientes: usuario, senha, atividade e aplicativo sao obrigatorios");
		}

		return data;
	}

	private static GPBotData getDataFromArgs(String[] args) {
		GPBotData data = new GPBotData();

		Stream.of(InputParameterEnum.values()).forEach(inputParameter -> {

			String argParameter = Stream.of(args).filter(arg -> arg.matches(inputParameter.getCommandLinePattern()))
					.findAny().orElse(null);

			if (StringUtils.isNotBlank(argParameter)) {
				String[] parameterParts = argParameter.split("=");
				String parameterValue = StringUtils.EMPTY;

				if (parameterParts.length > 1) {
					parameterValue = parameterParts[1].trim();
				}

				if (InputParameterEnum.ONLY_TODAY.equals(inputParameter)) {
					parameterValue = (StringUtils.isBlank(parameterValue)
							|| InputParameterEnum.ONLY_TODAY.matchesAnyExpectedValues(parameterValue)) ? "S" : "N";
				}

				data.set(inputParameter, parameterValue);
			}
		});

		return data;
	}

	private static GPBotData getDataFromDialog(GPBotData data) {

		if (data == null) {
			data = new GPBotData();
		}

		for (InputParameterEnum inputParameter : InputParameterEnum.values()) {

			if (!data.isSet(inputParameter)) {
				String parameterValue = requestInput(inputParameter, data);
				data.set(inputParameter, parameterValue.trim());
			}
		}

		return data;

	}

	private static String requestInput(InputParameterEnum inputParameter, GPBotData data) {
		String parameterInput = StringUtils.EMPTY;

		if (shouldRequestInput(inputParameter, data)) {
			System.out.print(inputParameter.getInputMessage());

			BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));

			try {

				if (InputParameterEnum.PASSWORD.equals(inputParameter)) {
					parameterInput = new String(System.console().readPassword());

				} else {
					parameterInput = bufferedReader.readLine();
				}

				if (InputParameterEnum.ONLY_TODAY.equals(inputParameter)) {
					parameterInput = InputParameterEnum.ONLY_TODAY.matchesAnyExpectedValues(parameterInput) ? "S" : "N";
				}

				return parameterInput;

			} catch (Exception e) {
				throw new IllegalArgumentException("Erro ao ler parametro de entrada", e);
			}
		}

		return parameterInput;
	}

	private static boolean shouldRequestInput(InputParameterEnum inputParameter, GPBotData data) {
		boolean parameterIsNowUseless = (Boolean.TRUE.equals(data.isTodayOnly())
				&& (InputParameterEnum.MONTH.equals(inputParameter)
						|| InputParameterEnum.CUSTOM_DAYS.equals(inputParameter)
						|| InputParameterEnum.SKIP_DAYS.equals(inputParameter)));

		return !parameterIsNowUseless;
	}

	public static ChromeDriver getDriver() {
		Map<String, String> chromeDriverVersionsByChrome = getChromeDriverVersionsByChrome();

		String chromeDrivePath = GPBotApplication.getPath() + "/lib/chromedriver/{0}/chromedriver";
		String chromeVersion = getChromeVersion();
		ChromeDriver driver = null;

		if (StringUtils.isNotBlank(chromeVersion)) {

			if (!chromeDriverVersionsByChrome.keySet().contains(chromeVersion)) {
				throw new UnsupportedOperationException("Versao do Chrome nao suportada");
			} else {
				LOGGER.info("Identificada versao " + chromeVersion + " do Chrome");
			}

			chromeDrivePath = MessageFormat.format(chromeDrivePath, chromeDriverVersionsByChrome.get(chromeVersion));
			System.setProperty("webdriver.chrome.driver", chromeDrivePath);

			driver = new ChromeDriver();

		} else {
			LOGGER.info("Versao do Chrome nao identificada, identificando versao do ChromeDriver aplicavel...");

			for (String chromeDriverVersion : getChromeDriverVersionsByPriority()) {
				chromeDrivePath = MessageFormat.format(chromeDrivePath, chromeDriverVersion);

				try {
					System.setProperty("webdriver.chrome.driver", chromeDrivePath);
					driver = new ChromeDriver();
					break;

				} catch (Exception e) {
					// driver nao encontrado ainda
				}
			}

		}

		if (driver == null) {
			throw new UnsupportedOperationException("Nenhuma versao do ChromeDriver disponivel pode ser aplicada");
		}

		driver.manage().timeouts().implicitlyWait(60, TimeUnit.SECONDS);
		return driver;
	}

	private static String getChromeVersion() {
		String chromeVersion = null;

		try {
			InputStream inputStream = Runtime.getRuntime().exec("google-chrome --version").getInputStream();
			String versionCommandOutput = new BufferedReader(new InputStreamReader(inputStream)).readLine();

			if (versionCommandOutput.matches(CHROME_VERSION_COMMAND_OUTPUT_PATTERN)) {
				chromeVersion = versionCommandOutput.split("Google Chrome ")[1].substring(0, 2);
			}

		} catch (Exception e) {
			// erro ao detectar versao do Chrome, tentara os drivers disponiveis
			// sequencialmente
		}

		return chromeVersion;
	}

	private static Map<String, String> getChromeDriverVersionsByChrome() {
		Map<String, String> chromeDriverVersionsByChrome = new HashMap<>();
		chromeDriverVersionsByChrome.put("61", "2.34");
		chromeDriverVersionsByChrome.put("62", "2.34");
		chromeDriverVersionsByChrome.put("63", "2.34");
		chromeDriverVersionsByChrome.put("64", "2.37");
		chromeDriverVersionsByChrome.put("65", "2.37");
		chromeDriverVersionsByChrome.put("66", "2.37");
		chromeDriverVersionsByChrome.put("67", "2.38");
		return chromeDriverVersionsByChrome;
	}

	private static List<String> getChromeDriverVersionsByPriority() {
		return Arrays.asList("2.34", "2.37", "2.38");
	}
}
